lib/recursive-methods.rb

```diff
@@ -1,49 +1,198 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
+# https://github.com/stupendousC/recursion-writing
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if (n == 1) || (n == 0)
+    return 1
+  elsif n > 1
+    return n * factorial(n-1)
+  else
+    raise ArgumentError, "argument must be >= 0"
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  # accepts a string and returns the reverse of the string by reversing all letters and all words in the string
+  # ex: reverse("hello world") will convert the input string to "dlrow olleh"
+  if s.length <= 1
+    return s 
+  else
+    return s[-1] + reverse(s[1...-1]) + s[0]
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.